### DISCUSSION TITLE
Best practice for unit testing routes that depend on an authenticated user

### DISCUSSION THREAD

**Discussion by Haydnspass** [fastapi-users/fastapi-users#677] [Q&A]
Hey there,

I would like to unit test all of my endpoints, say I got one where a User with attr 'mod' (but not a superuser) can query users by id.

```python
from fastapi_users import models

class User(models.BaseUser):
    is_mod: Optional[bool] = False


# custom attrs cannot be verified via current_user directly
mod = fastapi_users.current_user(
    active=True,
    verified=True,
)

@app.get("/mod/user/{id}", response_model=User)
async def read_user_by_id(uid: UUID4, user_authenticated: model.User = Depends(mod)):
    if not user_authenticated.is_mod:
        raise HTTPException(401)

    user_by_id = await user_db.get(uid)
    return user_by_id
```

how could I test this properly? My first try (looking at the fastapi docs) has been

```python
# imports
client = TestClient(main.app)

# tries to mock current user
async def override_mod_dep():
    return model.User(
        id="59408629-e9a1-47f9-97c6-545f8013290a",
        active=True,
        verified=True,
        mod=False
    )


def test_read_user_by_id():
    # not authenticated can not access
    response = client.get("/mod/user/62591021-f808-44b1-8345-6572c6b3e4d6")
    assert response.status_code == 401

    # authenticated but not a mod
    main.app.dependency_overrides[app.requester] = override_requester_dep
    response = client.get("/mod/user/62591021-f808-44b1-8345-6572c6b3e4d6")
    assert response.status_code == 401

    main.app.dependency_overrides = {}
```
unfortunately this fails with 422: "b'{"detail":[{"loc":["query","uid"],"msg":"field required","type":"value_error.missing"}]}'"
I think I am close, but I am missing something.
Are there general best practices how to test something like this?

**Comment by frankie567** [comment]
Hi there,

Sorry for the delayed response 🙂

The problem you have here is because you didn't use the same parameter name in the route pattern (`id`) and the function parameter (`uid`). This is why you get a 422, because FastAPI can't find the corresponding value. This should be:

```py
@app.get("/mod/user/{id}", response_model=User)
async def read_user_by_id(id: UUID4, user_authenticated: model.User = Depends(mod)):
    if not user_authenticated.is_mod:
        raise HTTPException(401)

    user_by_id = await user_db.get(uid)
    return user_by_id
```

Besides, I don't think you are overriding the right dependency in your test. What is `app.requester`? If you call your authentication dependency `mod`, you should override `mod`:

```py
 main.app.dependency_overrides[mod] = override_requester_dep
```

That said, I invite you to follow #671, because there may be some problems with that that we are currently discussing.
